File: backend/app/services/insight_sections.py
"""Insights page section preferences (#821) + Phase 6 versioned hub shrink (D5).

Configurable order and visibility for the ``/insights`` page blocks, mirroring
the Home screen mechanism (#584). The main insight feed (``insight_feed``) is a
locked section: always enabled (never hidden), but freely reorderable.

Phase 6 introduces ``CURRENT_INSIGHT_SECTIONS_VERSION`` so a one-time migrate
can shrink the Layer-1 default without wiping conscious Settings choices.
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import Literal

from app.services.section_preferences import (
    SectionPreference,
    merge_sections,
    normalize_sections,
)

InsightSectionKey = Literal[
    "stage_header",
    "correlation_matrix",
    "insight_feed",
    "lag_heatmap",
    "dismissed",
    "symptom_analytics",
    "tag_groups",
    "tag_cooccurrence",
]

VALID_INSIGHT_SECTION_KEYS: frozenset[str] = frozenset(
    {
        "stage_header",
        "correlation_matrix",
        "insight_feed",
        "lag_heatmap",
        "dismissed",
        "symptom_analytics",
        "tag_groups",
        "tag_cooccurrence",
    }
)

# The main feed cannot be hidden — only reordered.
LOCKED_INSIGHT_SECTION_KEYS: frozenset[str] = frozenset({"insight_feed"})

InsightSectionPreference = SectionPreference

# Pre-Phase-6 default (all on). Used only to detect "saved defaults" layouts.
LEGACY_DEFAULT_INSIGHT_SECTIONS: list[SectionPreference] = [
    {"key": "stage_header", "enabled": True},
    {"key": "correlation_matrix", "enabled": True},
    {"key": "insight_feed", "enabled": True},
    {"key": "lag_heatmap", "enabled": True},
    {"key": "dismissed", "enabled": True},
    {"key": "symptom_analytics", "enabled": True},
    {"key": "tag_groups", "enabled": True},
    {"key": "tag_cooccurrence", "enabled": True},
]

# Phase 6 / D5: slim Layer-1 entry. Optional blocks stay valid for Settings.
DEFAULT_INSIGHT_SECTIONS: list[SectionPreference] = [
    {"key": "stage_header", "enabled": True},
    {"key": "insight_feed", "enabled": True},
    {"key": "correlation_matrix", "enabled": False},
    {"key": "lag_heatmap", "enabled": False},
    {"key": "dismissed", "enabled": False},
    {"key": "symptom_analytics", "enabled": False},
    {"key": "tag_groups", "enabled": False},
    {"key": "tag_cooccurrence", "enabled": False},
]

# Keys that leave the default viewport in v2 (still in validKeys).
SHRINK_INSIGHT_SECTION_KEYS: frozenset[str] = frozenset(
    {
        "correlation_matrix",
        "lag_heatmap",
        "dismissed",
        "symptom_analytics",
        "tag_groups",
        "tag_cooccurrence",
    }
)

# Version 1 = legacy all-on layout era; version 2 = slim hub defaults.
CURRENT_INSIGHT_SECTIONS_VERSION = 2
LEGACY_INSIGHT_SECTIONS_VERSION = 1


def merge_insight_sections(
    stored: Sequence[object] | None,
) -> list[SectionPreference]:
    """Resolve stored preferences with defaults for missing or unknown keys."""
    return merge_sections(
        stored,
        defaults=DEFAULT_INSIGHT_SECTIONS,
        valid_keys=VALID_INSIGHT_SECTION_KEYS,
        locked_keys=LOCKED_INSIGHT_SECTION_KEYS,
    )


def normalize_insight_sections(
    sections: Sequence[object] | None,
) -> list[SectionPreference] | None:
    """Validate and dedupe a PATCH payload; empty list is allowed.

    Locked keys (``insight_feed``) are forced to ``enabled=True``.
    """
    return normalize_sections(
        sections,
        valid_keys=VALID_INSIGHT_SECTION_KEYS,
        locked_keys=LOCKED_INSIGHT_SECTION_KEYS,
    )


def _is_exact_legacy_layout(stored: Sequence[object]) -> bool:
    """Compare the raw stored array before normalization changes locked flags."""
    return len(stored) == len(LEGACY_DEFAULT_INSIGHT_SECTIONS) and all(
        isinstance(raw, dict)
        and raw.get("key") == default["key"]
        and raw.get("enabled") is default["enabled"]
        for raw, default in zip(stored, LEGACY_DEFAULT_INSIGHT_SECTIONS, strict=True)
    )


def migrate_insight_sections_to_current(
    stored: Sequence[object] | None,
    *,
    version: int | None,
) -> tuple[list[SectionPreference] | None, int, bool]:
    """One-time v1→v2 hub shrink.

    Returns ``(sections_to_persist_or_None, new_version, dirty)``.
    ``None`` sections means leave the JSON column NULL (use defaults via merge).
    """
    current_version = version if version is not None else LEGACY_INSIGHT_SECTIONS_VERSION
    if current_version >= CURRENT_INSIGHT_SECTIONS_VERSION:
        normalized = normalize_insight_sections(stored)
        return normalized, current_version, False

    if not stored:
        # NULL/empty → new slim defaults via merge; persist version only.
        return None, CURRENT_INSIGHT_SECTIONS_VERSION, True

    normalized = normalize_insight_sections(stored) or []
    if not normalized:
        return None, CURRENT_INSIGHT_SECTIONS_VERSION, True

    if _is_exact_legacy_layout(stored):
        # Only the complete, ordered, unchanged v1 template is known to be inherited.
        return (
            [section.copy() for section in DEFAULT_INSIGHT_SECTIONS],
            CURRENT_INSIGHT_SECTIONS_VERSION,
            True,
        )

    # A partial, reordered, or changed layout is user state. Keep every explicit
    # flag and its order; append missing known keys using today's defaults.
    return merge_insight_sections(normalized), CURRENT_INSIGHT_SECTIONS_VERSION, True

File: apps/web/src/lib/utils/insightSections.ts
import type { InsightSectionKey, InsightSectionPreference } from '$lib/api/preferences';
import { createSectionUtils } from '$lib/utils/sectionPreferences';

export type { InsightSectionKey, InsightSectionPreference };

/** The main feed can never be hidden — only reordered. */
export const LOCKED_INSIGHT_SECTION_KEYS: InsightSectionKey[] = ['insight_feed'];

/** Pre-Phase-6 all-on layout — used only to detect "saved defaults". */
export const LEGACY_DEFAULT_INSIGHT_SECTIONS: InsightSectionPreference[] = [
  { key: 'stage_header', enabled: true },
  { key: 'correlation_matrix', enabled: true },
  { key: 'insight_feed', enabled: true },
  { key: 'lag_heatmap', enabled: true },
  { key: 'dismissed', enabled: true },
  { key: 'symptom_analytics', enabled: true },
  { key: 'tag_groups', enabled: true },
  { key: 'tag_cooccurrence', enabled: true },
];

/**
 * Phase 6 / D5 slim Layer-1 default. Optional blocks stay in validKeys so
 * Settings can re-enable them; they are no longer in the entry viewport.
 */
export const DEFAULT_INSIGHT_SECTIONS: InsightSectionPreference[] = [
  { key: 'stage_header', enabled: true },
  { key: 'insight_feed', enabled: true },
  { key: 'correlation_matrix', enabled: false },
  { key: 'lag_heatmap', enabled: false },
  { key: 'dismissed', enabled: false },
  { key: 'symptom_analytics', enabled: false },
  { key: 'tag_groups', enabled: false },
  { key: 'tag_cooccurrence', enabled: false },
];

export const SHRINK_INSIGHT_SECTION_KEYS: InsightSectionKey[] = [
  'correlation_matrix',
  'lag_heatmap',
  'dismissed',
  'symptom_analytics',
  'tag_groups',
  'tag_cooccurrence',
];

export const CURRENT_INSIGHT_SECTIONS_VERSION = 2;
export const LEGACY_INSIGHT_SECTIONS_VERSION = 1;

const VALID_INSIGHT_SECTION_KEYS: InsightSectionKey[] = [
  'stage_header',
  'correlation_matrix',
  'insight_feed',
  'lag_heatmap',
  'dismissed',
  'symptom_analytics',
  'tag_groups',
  'tag_cooccurrence',
];

const insightSectionUtils = createSectionUtils<InsightSectionKey>({
  validKeys: VALID_INSIGHT_SECTION_KEYS,
  defaults: DEFAULT_INSIGHT_SECTIONS,
  lockedKeys: LOCKED_INSIGHT_SECTION_KEYS,
});

export function mergeInsightSections(
  stored: InsightSectionPreference[] | null | undefined
): InsightSectionPreference[] {
  return insightSectionUtils.merge(stored);
}

export function resolveEnabledInsightSections(
  sections: InsightSectionPreference[]
): InsightSectionPreference[] {
  return insightSectionUtils.resolveEnabled(sections);
}

export function normalizeInsightSectionsForSave(
  sections: InsightSectionPreference[]
): InsightSectionPreference[] {
  return insightSectionUtils.normalizeForSave(sections);
}

export function isInsightSectionLocked(key: InsightSectionKey): boolean {
  return insightSectionUtils.isLocked(key);
}

function sameLayout(
  left: readonly InsightSectionPreference[],
  right: readonly InsightSectionPreference[]
): boolean {
  if (left.length !== right.length) return false;
  return left.every(
    (section, index) =>
      section.key === right[index]?.key && section.enabled === right[index]?.enabled
  );
}

/**
 * Client-side mirror of the backend v1→v2 migrate (offline / tests).
 * Prefer trusting GET /user/preferences after the server has persisted.
 */
export function migrateInsightSectionsToCurrent(
  stored: InsightSectionPreference[] | null | undefined,
  version: number | null | undefined
): { sections: InsightSectionPreference[] | null; version: number; dirty: boolean } {
  const currentVersion = version ?? LEGACY_INSIGHT_SECTIONS_VERSION;
  if (currentVersion >= CURRENT_INSIGHT_SECTIONS_VERSION) {
    return {
      sections: stored?.length ? normalizeInsightSectionsForSave(stored) : null,
      version: currentVersion,
      dirty: false,
    };
  }

  if (!stored?.length) {
    return { sections: null, version: CURRENT_INSIGHT_SECTIONS_VERSION, dirty: true };
  }

  const normalized = normalizeInsightSectionsForSave(stored);
  if (sameLayout(stored, LEGACY_DEFAULT_INSIGHT_SECTIONS)) {
    return {
      sections: DEFAULT_INSIGHT_SECTIONS.map((section) => ({ ...section })),
      version: CURRENT_INSIGHT_SECTIONS_VERSION,
      dirty: true,
    };
  }

  return {
    // Preserve explicit order and flags. Missing known keys get current defaults.
    sections: insightSectionUtils.merge(normalized),
    version: CURRENT_INSIGHT_SECTIONS_VERSION,
    dirty: true,
  };
}

/** Optional hub tools that live behind Settings when off by default. */
export const INSIGHT_TOOL_SECTION_KEYS: InsightSectionKey[] = [
  'symptom_analytics',
  'tag_groups',
  'tag_cooccurrence',
  'lag_heatmap',
  'correlation_matrix',
];
